solutions/platinum/cses-1160.mdx

```diff
@@ -199,4 +199,88 @@ int main() {
 ```
 
 </CPPSection>
+<JavaSection>
+
+```java
+
+import java.io.*;
+import java.util.StringTokenizer;
+
+public class PlanetQueries2 {
+	static Kattio x = new Kattio();
+	static PrintWriter pw = new PrintWriter(System.out);
+	static int n;
+	static int q;
+	static int[] arr;
+	static int[][] binaryLifting;
+	static int[] len;
+
+	public static void main(String[] args) {
+		n = x.nextInt();
+		q = x.nextInt();
+		arr = new int[n + 1];
+		for (int i = 1; i <= n; i++) { arr[i] = x.nextInt(); }
+		binaryLifting = new int[n + 1][20];
+		len = new int[n + 1];
+		/*
+		 * in preprocess, doing dfs, for each node, to ensure to covering
+		 * more than one connected component
+		 */
+		for (int i = 1; i <= n; i++) { binaryLifting[i][0] = arr[i]; }
+		// visited array ensures to cover more than one connected components
+		boolean visited[] = new boolean[n + 1];
+		for (int i = 1; i <= n; i++) {
+			if (!visited[i]) { dfs(i, visited); }
+		}
+		StringBuilder output = new StringBuilder();
```

Comment:
do you use this variable?

Reply:
nope, was using it for debugging, removing it.